File: python/2020/12/adv.py
import math


def get_input() -> str:
    with open("input.txt") as f:
        return f.read()


def get_test_input() -> str:
    with open("test.txt") as f:
        return f.read()


class Instruction:
    def __init__(self, command: str, num: int) -> None:
        self.command = command
        self.num = num

    @classmethod
    def from_string(cls, line: str) -> "Instruction":
        command = line[0]
        num = int(line[1:])

        return Instruction(command, num)

    def __repr__(self) -> str:
        return f"command: {self.command}, num: {self.num}"


class StateDirection:
    def __init__(self) -> None:
        self.pos_x = 0
        self.pos_y = 0
        self.direction = "E"

    def __repr__(self) -> str:
        return f"Position ({self.pos_x}, {self.pos_y}), Direction: {self.direction}"

    def turn_left(self):
        if self.direction == "E":
            self.direction = "N"
        elif self.direction == "N":
            self.direction = "W"
        elif self.direction == "W":
            self.direction = "S"
        elif self.direction == "S":
            self.direction = "E"
        else:
            raise Exception(f"Invalid direction: {self.direction}")

    def turn_right(self):
        if self.direction == "E":
            self.direction = "S"
        elif self.direction == "S":
            self.direction = "W"
        elif self.direction == "W":
            self.direction = "N"
        elif self.direction == "N":
            self.direction = "E"
        else:
            raise Exception(f"Invalid direction: {self.direction}")

    def apply_instruction(self, inst: Instruction):
        if inst.command == "N":
            self.pos_y += inst.num
        elif inst.command == "S":
            self.pos_y -= inst.num
        elif inst.command == "E":
            self.pos_x += inst.num
        elif inst.command == "W":
            self.pos_x -= inst.num
        elif inst.command == "L":
            count = int(inst.num / 90)
            for _ in range(count):
                self.turn_left()
        elif inst.command == "R":
            count = int(inst.num / 90)
            for _ in range(count):
                self.turn_right()
        elif inst.command == "F":
            new_inst = Instruction(self.direction, inst.num)
            self.apply_instruction(new_inst)
        else:
            raise Exception(f"Invalid instruction command: {inst.command}")


class StateWaypoint:
    def __init__(self) -> None:
        self.pos_x = 0
        self.pos_y = 0
        self.way_x = 10
        self.way_y = 1

    def __repr__(self) -> str:
        return f"Position ({self.pos_x}, {self.pos_y}), Waypoint ({self.way_x}, {self.way_y})"

    def rotate(self, angle: int):
        angle_rad = math.radians(angle)
        sin = math.sin(angle_rad)
        cos = math.cos(angle_rad)
        x = self.way_x * cos - self.way_y * sin
        y = self.way_x * sin + self.way_y * cos
        self.way_x = round(x)
        self.way_y = round(y)

    def apply_instruction(self, inst: Instruction):
        if inst.command == "N":
            self.way_y += inst.num
        elif inst.command == "S":
            self.way_y -= inst.num
        elif inst.command == "E":
            self.way_x += inst.num
        elif inst.command == "W":
            self.way_x -= inst.num
        elif inst.command == "L":
            self.rotate(inst.num)
        elif inst.command == "R":
            self.rotate(-inst.num)
        elif inst.command == "F":
            self.pos_x += self.way_x * inst.num
            self.pos_y += self.way_y * inst.num
        else:
            raise Exception(f"Invalid instruction command: {inst.command}")


def get_distance_direction(input: str) -> int:
    insts = [Instruction.from_string(line) for line in input.splitlines()]
    state = StateDirection()
    for inst in insts:
        state.apply_instruction(inst)

    return abs(state.pos_x) + abs(state.pos_y)


def get_distance_waypoint(input: str) -> int:
    insts = [Instruction.from_string(line) for line in input.splitlines()]
    state = StateWaypoint()
    for inst in insts:
        state.apply_instruction(inst)

    return abs(state.pos_x) + abs(state.pos_y)


def part_1(input: str):
    answer_1 = get_distance_direction(input)

    print(f"Part 1 answer is {answer_1}")


def part_2(input: str):
    answer_2 = get_distance_waypoint(input)

    print(f"Part 2 answer is {answer_2}")


def run_test():
    input = get_test_input()
    answer_1 = get_distance_direction(input)
    assert answer_1 == 25, f"Expected: '{25}', Found: '{answer_1}'"
    answer_2 = get_distance_waypoint(input)
    assert answer_2 == 286, f"Expected: '{287}', Found: '{answer_2}'"

    print("Tests passed!")


if __name__ == "__main__":
    run_test()
    input = get_input()
    part_1(input)
    part_2(input)

File: ruby/2020/12/adv.rb
# frozen_string_literal: true

require '../../aoc_base'

# Year 2020 Day 12
class Aoc12 < AocBase
  # Represent a ship instruction
  class Instruction
    attr_reader :command, :num

    def initialize(command, num)
      @command = command
      @num = num
    end

    def self.from_string(text)
      command = text[0]
      num = text[1..].to_i
      Instruction.new(command, num)
    end

    def to_s
      "Command: #{@command}, Num: #{@num}"
    end

    def inspect
      to_s
    end
  end

  # The ship state using direction
  class StateDirection
    DIR_LEFT = { 'E' => 'N', 'N' => 'W', 'W' => 'S', 'S' => 'E' }.freeze
    DIR_RIGHT = { 'E' => 'S', 'S' => 'W', 'W' => 'N', 'N' => 'E' }.freeze
    def initialize
      @pos_x = 0
      @pos_y = 0
      @direction = 'E'
    end

    def manhatten_distance
      @pos_x.abs + @pos_y.abs
    end

    def turn_left
      @direction = DIR_LEFT[@direction]
    end

    def turn_right
      @direction = DIR_RIGHT[@direction]
    end

    def apply_inst(inst)
      case inst.command
      when 'N'
        @pos_y += inst.num
      when 'S'
        @pos_y -= inst.num
      when 'E'
        @pos_x += inst.num
      when 'W'
        @pos_x -= inst.num
      when 'L'
        count = inst.num / 90
        count.times do |_|
          turn_left
        end
      when 'R'
        count = inst.num / 90
        count.times do |_|
          turn_right
        end
      when 'F'
        new_inst = Instruction.new(@direction, inst.num)
        apply_inst(new_inst)
      else
        raise "Invalid command #{inst.command}"
      end
    end
  end

  class StateWaypoint
    def initialize
      @pos_x = 0
      @pos_y = 0
      @way_x = 10
      @way_y = 1
    end

    def manhatten_distance
      @pos_x.abs + @pos_y.abs
    end

    def rotate(angle_degree)
      angle_rad = angle_degree * Math::PI / 180
      sin = Math.sin(angle_rad)
      cos = Math.cos(angle_rad)
      x = @way_x * cos - @way_y * sin
      y = @way_x * sin + @way_y * cos
      @way_x = x.round
      @way_y = y.round
    end

    def move_forward(num)
      @pos_x += @way_x * num
      @pos_y += @way_y * num
    end

    def apply_inst(inst)
      case inst.command
      when 'N'
        @way_y += inst.num
      when 'S'
        @way_y -= inst.num
      when 'E'
        @way_x += inst.num
      when 'W'
        @way_x -= inst.num
      when 'L'
        rotate(inst.num)
      when 'R'
        rotate(-inst.num)
      when 'F'
        move_forward(inst.num)
      else
        raise "Invalid command #{inst.command}"
      end
    end
  end

  def calc_distance(input, state)
    input
      .lines
      .map { |line| Instruction.from_string(line) }
      .each do |inst|
        state.apply_inst(inst)
      end

    state.manhatten_distance
  end

  def calc_direction_distance(input)
    state = StateDirection.new
    calc_distance(input, state)
  end

  def calc_waypoint_distance(input)
    state = StateWaypoint.new
    calc_distance(input, state)
  end

  def part_one
    calc_direction_distance(@input)
  end

  def part_two
    calc_waypoint_distance(@input)
  end

  def do_tests
    assert_equal calc_direction_distance(@test_input), 25
    assert_equal calc_waypoint_distance(@test_input), 286
    puts 'tests pass'
  end
end

if __FILE__ == $PROGRAM_NAME
  aoc = Aoc12.new
  aoc.run
end
